vyper/venom/passes/dft.py

```diff
@@ -1,81 +1,152 @@
+
+    def _process_basic_block(self, bb: IRBasicBlock) -> None:
+        self.function.append_basic_block(bb)
+
+        self._calculate_dependency_graphs(bb)
+        self.instructions = list(bb.pseudo_instructions)
+
+        # Compute the number of instructions that are dependent on each instruction
+        self.visited_instructions = OrderedSet()
+        non_phi_instructions = list(bb.non_phi_instructions)
+        for inst in non_phi_instructions:
+            self._calculate_instruction_offspring_count_r(inst)
+
+        # Compute entry points in the graph of instruction dependencies
+        entry_instructions: OrderedSet[IRInstruction] = OrderedSet(non_phi_instructions)
+        for inst in non_phi_instructions:
+            to_remove = self.ida.get(inst, OrderedSet())
+            if len(to_remove) > 0:
+                entry_instructions.dropmany(to_remove)
+
+        # Move the terminator instruction to the end of the list
+        terminator = next(inst for inst in entry_instructions if inst.is_bb_terminator)
+        assert terminator is not None, f"Basic block should have a terminator instruction {bb}"
+        entry_instructions.remove(terminator)
+        entry_instructions.add(terminator)
+
+        self.visited_instructions = OrderedSet()
+        for inst in entry_instructions:
+            self._process_instruction_r(self.instructions, inst)
 
-                # if the instruction is a terminator, we need to place
-                # it at the end of the basic block
-                # along with all the instructions that "lead" to it
-                self._process_instruction_r(bb, uses_this, offset)
+        bb.instructions = self.instructions
+        assert bb.is_terminated, f"Basic block should be terminated {bb}"
 
+    def _process_instruction_r(self, instructions: list[IRInstruction], inst: IRInstruction):
         if inst in self.visited_instructions:
             return
         self.visited_instructions.add(inst)
-        self.inst_order_num += 1
 
-        if inst.is_bb_terminator:
-            offset = len(bb.instructions)
-
-        if inst.opcode == "phi":
-            # phi instructions stay at the beginning of the basic block
-            # and no input processing is needed
-            # bb.instructions.append(inst)
-            self.inst_order[inst] = 0
+        if inst.is_pseudo:
             return
 
-        for op in inst.get_input_variables():
-            target = self.dfg.get_producing_instruction(op)
-            assert target is not None, f"no producing instruction for {op}"
-            if target.parent != inst.parent or target.fence_id != inst.fence_id:
-                # don't reorder across basic block or fence boundaries
+        children = sorted(
+            self.ida[inst],
+            key=lambda x: (
+                -self.inst_offspring_count[x] + (x.opcode == "iszero") * 10,
+                inst.operands.index(x.output) if x.output in inst.operands else 0,
+            ),
+        )
```

Comment:
the following seems to do slightly better:
```suggestion
        children = sorted(
            self.ida[inst],
            key=lambda x: (
                inst.operands.index(x.output) if x.output in inst.operands else 0,
                -self.inst_offspring_count[x],
            ),
        )
```
i guess inst_offspring_count is too coarse for what we want. i think the weight should be more like, recursively computing `inst.operands.index(x.output) if x.output in inst.operands else 0`, through the whole subgraph.

Reply:
Maybe this file will be clearer if the operand dependencies and barriers are separated into two data structures